pkgs/build-support/make-darwin-app/default.nix

```diff
@@ -0,0 +1,74 @@
+{
+  lib,
+  stdenv,
+  makeWrapper,
+  undmg,
+}:
+
+# Returns the makeDarwinApp function
+{
+  /** ${appName}.app */
+  appName,
+  /** Name of the file found in ${appName}.app/Contents/MacOS/${binaryName}
+    * to be executed
+    * TODO: detect this automatically from the .plist file?
+    */
+  binaryName,
+  pname,
+  version,
+  /** a nix path or derivation to be used as mkDerivation.src */
+  src,
+  /** is src a .dmg .
+    * These contain multiple folders after extraction
+    * which requires special treatment in the unpackPhase
+    */
+  isDmg ? false,
+  /** Whether to use makeWrapper for the binary.
+    * For some packages symbolic links just don't work
+    */
+  wrapBinary ? false,
+  /** The meta attribute passed to mkDerivation */
+  packageMeta ? {},
+  /** Can use a custom function e.g stdenvNoCC.mkDerivation */
+  derivationFunction ? stdenv.mkDerivation,
+}:
+derivationFunction (
+    # .dmg extracts have multiple folders
+    # the unpack phase requires only one folder
+    (lib.attrsets.optionalAttrs isDmg {
+      sourceRoot = "${appName}.app";
+    }) // {
+    inherit pname version src;
+    nativeBuildInputs = (lib.optional wrapBinary makeWrapper) ++ (lib.optional isDmg undmg) ;
+
+    installPhase = ''
+    appDir="$out/Applications/${appName}.app"
+    binDir="$appDir"/Contents/MacOS
+
+    mkdir -p $out/bin
+    if [[ "${toString isDmg}" = "1" ]] ; then
+      # .dmg uses sourceRoot, which means PWD is already in the sourceRoot
+      mkdir -p $appDir
+      cp -r Contents $appDir
+    else
+      # expects a .app folder in the source folder
+      mkdir $out/Applications
+      cp -r ${appName}.app/ $appDir
+    fi
+
+    # Make application available in bin/
+    if [[ "${toString wrapBinary}" = "1" ]] ; then
+      makeWrapper $binDir/${binaryName} $out/bin/${pname}
```

Comment:
Do we have any idea why? Is this about the binary not having execute permissions or something?

Reply:
It seems to have something to do with the location of the binary. It can't find the framework and libs otherwise. The same goes for firefox, which comes from a `.dmg`. Executing the wrapped binary from any CWD works, but if the binary is moved or linked to (`ln`), it doesn't work.

If someone else wants to invest the time to patch the binary, they can be my guest.